src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java

```diff
@@ -62,4 +105,97 @@ private static int setInterval(FabricClientCommandSource source, int interval) t
+        }
+
+        @Override
+        public ItemInput parse(StringReader reader) throws CommandSyntaxException {
+            return lastItem = super.parse(reader);
+        }
+    }
+
+    private static Combined<Predicate<ItemStack>, String> createPredicate(CommandContext<FabricClientCommandSource> context, String argName) {
+        try {
+            Result<Combined<ItemInput, Combined<Integer, Integer>>> result = getWithString(context, argName, null);
+            var combined = result.value();
+            var item = combined.first().getItem();
+            var min = combined.second().first();
+            var max = combined.second().second();
+            return new Combined<>((stack) -> stack.is(item)
+                && ((min > item.getDefaultMaxStackSize() || min <= stack.getCount())
+                && (max > item.getDefaultMaxStackSize() || stack.getCount() <= max)),
+                result.string().replaceAll("\\* \\*","*").replaceAll("([\\d*]+) ([\\d*]+)$", "$1-$2"));
+        } catch (IllegalArgumentException ignored) { }
+        return null;
+    }
+
+    private static int setGoal(CommandContext<FabricClientCommandSource> context) {
+        CCrackVillager.Offer offer = new CCrackVillager.Offer();
+
+        var combined = createPredicate(context, "firstitem");
+        if(combined != null) offer.withFirst(combined.first(), combined.second());
+        combined = createPredicate(context, "seconditem");
+        if(combined != null) offer.withSecond(combined.first(), combined.second());
+        combined = createPredicate(context, "resultitem");
+        if(combined != null) offer.withResult(combined.first(), combined.second());
+
+        try {
+            Result<Combined<Enchantment, Integer>> result2 = getWithString(context, "enchantment", null);
+            var enchantment = result2.value().first();
+            offer.andEnchantment((stack) -> {
+                var enchantments = EnchantmentHelper.getEnchantmentsForCrafting(stack);
+                var level = enchantments.getLevel(enchantment);
+                return (result2.value().second() > enchantment.getMaxLevel() && level > 0) || result2.value().second() == level;
+            }, result2.string());
+        } catch (IllegalArgumentException ignored) { }
+
+        context.getSource().sendFeedback(Component.literal((CCrackVillager.goalOffers.size()+1) + ": " + offer));
+
+        CCrackVillager.goalOffers.add(offer);
+
+        return Command.SINGLE_SUCCESS;
+    }
+
+
+
+    static LiteralArgumentBuilder<FabricClientCommandSource> genFirst(CommandBuildContext context) {
+        var item = item(context);
+        Supplier<Integer> supplier = () -> item.lastItem.getItem().getDefaultMaxStackSize();
+        return literal("first")
+            .then(argument("firstitem", withString(combined(item, combined(integer(1,supplier).allowAny(), integer(1,supplier).allowAny()))))
+                .executes(CrackVillagerRNGCommand::setGoal)
```

Comment:
wait, so a valid goal can be a first item with no specific result or anything else set?